ui/ozone/platform/starboard/platform_event_source_starboard.cc

```diff
@@ -0,0 +1,175 @@
+// Copyright 2024 The Cobalt Authors. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "ui/ozone/platform/starboard/platform_event_source_starboard.h"
+
+#include "base/logging.h"
+#include "starboard/event.h"
+#include "starboard/input.h"
+#include "starboard/key.h"
+#include "starboard/log.h"
+#include "ui/events/event.h"
+
+#include "base/containers/fixed_flat_map.h"
+#include "ui/events/keycodes/dom/dom_code.h"
+#include "ui/events/keycodes/dom/dom_key.h"
+#include "ui/events/keycodes/keyboard_code_conversion.h"
+
+#include "ui/events/types/event_type.h"
+
+namespace starboard {
+namespace {
+
+constexpr auto kSbKeyToDomCodeMap = base::MakeFixedFlatMap<SbKey, ui::DomCode>({
+    {kSbKeySpace, ui::DomCode::MEDIA_PLAY_PAUSE},
+
+    {kSbKeyReturn, ui::DomCode::ENTER},
+    {kSbKeySelect, ui::DomCode::SELECT},
+    {kSbKeyUp, ui::DomCode::ARROW_UP},
+    {kSbKeyDown, ui::DomCode::ARROW_DOWN},
+    {kSbKeyLeft, ui::DomCode::ARROW_LEFT},
+    {kSbKeyRight, ui::DomCode::ARROW_RIGHT},
+    {kSbKeyBack, ui::DomCode::BROWSER_BACK},
+
+    {kSbKeyMediaPlayPause, ui::DomCode::MEDIA_PLAY_PAUSE},
+    {kSbKeyMediaRewind, ui::DomCode::MEDIA_REWIND},
+    {kSbKeyMediaFastForward, ui::DomCode::MEDIA_FAST_FORWARD},
+    {kSbKeyMediaNextTrack, ui::DomCode::MEDIA_TRACK_NEXT},
+    {kSbKeyMediaPrevTrack, ui::DomCode::MEDIA_TRACK_PREVIOUS},
+    {kSbKeyPause, ui::DomCode::MEDIA_PAUSE},
+    {kSbKeyPlay, ui::DomCode::MEDIA_PLAY},
+    {kSbKeyMediaStop, ui::DomCode::MEDIA_STOP},
+
+    {kSbKeyMenu, ui::DomCode::HOME},
+    {kSbKeyChannelUp, ui::DomCode::CHANNEL_UP},
+    {kSbKeyChannelDown, ui::DomCode::CHANNEL_DOWN},
+    {kSbKeyClosedCaption, ui::DomCode::CLOSED_CAPTION_TOGGLE},
+#if SB_API_VERSION >= 15
+    {kSbKeyRecord, ui::DomCode::MEDIA_RECORD},
+#endif  // SB_API_VERSION >=15
+});
+
+// TODO(tholcombe): replace this with SequencedTaskRunner. See
+// chromecast/starboard/chromecast/events/starboard_event_source.cc for
+// implementation reference.
+static scoped_refptr<base::TaskRunner> g_reply_runner_;
```

Comment:
This is not thread-safe and is used by both the Starboard thread and the Chromium threads.